// gamsl_server/engine.js
/**
 * Created by Gamer on 8/11/2015.
 */

module.exports = (function(){

    console.log("Creating engine module");

    var players = {};
    var npcs = {};

    npcs["npc1"] = new THREE.Vector3(0,1,0);

    npcs["npc2"] = new THREE.Vector3(5,1,0);

    npcs["npc3"] = new THREE.Vector3(10,1,0);

    npcs["npc4"] = new THREE.Vector3(15,1,0);

    for( var name in npcs ) npcs[name].speed = 15;

    function getState () {

        return {players:players,npcs:npcs}

    }

    function updatePlayer ( message ) {

        players[message.playerId] = message;

    }


    function tick ( elapsed ) {

        var dt = elapsed/1000;

        for( var name in npcs ) {

            npcs[name].x += npcs[name].speed*dt;
            if(npcs[name].x > 75 || npcs[name].x < -75){

                npcs[name].speed *= -1;

            }

        }

    }

    function addPlayer( message ) {

         players[message.playerId] = message;

    }

    function removePlayer ( id ) {

         delete players[ id ];

    }

    return {

        tick:tick,
        getState:getState,
        updatePlayer:updatePlayer,
        addPlayer:addPlayer,
        removePlayer:removePlayer

    }

})();
